### DISCUSSION TITLE
Unify sum type unwrapping with Option unwrapping

### DISCUSSION THREAD

**Discussion by ntrel** [vlang/v#11694] [Ideas]
In git we have:
* `if sum is T {/* sum is type T here */}` which shadows `sum` in the block.
* `if sum is T as var {` which safely unwraps `sum` with the name `var`.

Outside `if`, `sum is T` produces a boolean. What if `sum is T` produced an optional everywhere? If `sum` doesn't hold a `T` then `sum is T` would equal `none`. Then we could use optional handling with sum types:
```v
if val := sum is T {/* use val as T */}
val2 := sum is T or {...}
// use val2 as T
```
We could make `if opt {` valid with an optional value so that `if sum is T {` works naturally. That would be useful for shadow unwrapping an optional:
```v
fn f(val ?Type) {
    if val {
        val.foo() // val has type `Type`
    }
    else {
        println('no value')
    }
}
```
Shadowing seems at least as useful for safe unwrapping of an optional variable as it is for sum type instance unwrapping.

`opt or` could also work for shadow unwrapping an optional variable:
```v
val or {...}
// val here has type `Type`
```
All this unifies safe sum type unwrapping with optional handling in the language, for consistency.

**Comment by danieldaeschle** [comment]
> (because shadowing doesn't work if sum is a more complex expression)
What do you mean?

`is` is a simple operator which just checks the type and the compiler smartcasts the type where it can. You can do `sum is T as mysum`. Maybe that's what you want to achive?

**Comment by ntrel** [comment]
@danieldaeschle You can't use the safe unwrapping `if sum is T {` where sum of shadowed in the block if sum is actually a struct field, array indexing, or map key lookup expression.

**Comment by danieldaeschle** [comment]
`if sum.field is T as sum_field {` should work.

**Comment by ntrel** [comment]
@danieldaeschle: OK, thanks. I edited the summary in light of that.
